lib/acts_as_taggable_on/taggable/core.rb

```diff
@@ -49,8 +50,12 @@ def #{tag_type}_list
             def #{tag_type}_list=(new_tags)
               parsed_new_list = ActsAsTaggableOn.default_parser.new(new_tags).parse
 
-              if self.class.preserve_tag_order? || parsed_new_list.sort != #{tag_type}_list.sort
-                set_attribute_was('#{tag_type}_list', #{tag_type}_list)
+              if self.class.preserve_tag_order? || (parsed_new_list.sort != #{tag_type}_list.sort)
+                if ActsAsTaggableOn::Utils.legacy_activerecord?
+                  set_attribute_was("#{tag_type}_list", #{tag_type}_list)
+                else
+                  attribute_change("#{tag_type}_list")
```

Comment:
If my current understanding is correct, this line is supposed to set the 'previous' value of the *_list attribute. It seems that `attribute_change` is not the right method to do this. I’ve quickly looked into the code for `AttributeMutationTracker` but I’m not quite sure yet what’s the best way to replicate the behavior of `set_attribute_was` in AR 6.0.

Ideas welcome.

Reply:
Perhaps we could use `{attribute}_will_change!` as seen her in Rails 6 as well: https://api.rubyonrails.org/classes/ActiveModel/Dirty.html

_If an attribute is modified in-place then make use of [attribute_name]_will_change! to mark that the attribute is changing. Otherwise Active Model can't track changes to in-place attributes. Note that Active Record can detect in-place modifications automatically. You do not need to call [attribute_name]_will_change! on Active Record models._

Reply:
The problem is a bit different though. It’s that in this test case original value for `tag_list` is `nil` and not `["awesome", "epic"]` as expected. ActiveRecord picks up that the attribute has been changed, but it does not print the correct previous value.

```ruby
taggable = TaggableModel.find(@taggable.id)
taggable.tag_list
# => ["awesome", "epic"]
taggable.tag_list_will_change!
taggable.tag_list_change
# => [nil, nil]
taggable.tag_list = 'one'
taggable.tag_list_change
# => [nil, ["one"]]
```